src/clr-angular/utils/drag-and-drop/providers/drag-event-listener.ts

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import {Injectable, Renderer2} from "@angular/core";
+import {Observable} from "rxjs/Observable";
+import {Subject} from "rxjs/Subject";
+
+import {ClrDragDataTransfer, ClrDragEvent, ClrDragEventType} from "../interfaces/drag-event";
+import {ClrDragAndDropEventBus} from "./drag-and-drop-event-bus";
+
+@Injectable()
+export class ClrDragEventListener {
+    private draggabelEl: Node;
+
+    private listeners: [() => void];
+
+    private dragStart: Subject<ClrDragEvent> = new Subject<ClrDragEvent>();
+    private dragMove: Subject<ClrDragEvent> = new Subject<ClrDragEvent>();
+    private dragEnd: Subject<ClrDragEvent> = new Subject<ClrDragEvent>();
+
+    private hasDragStarted: boolean = false;
+
+    get dragStarted(): Observable<ClrDragEvent> {
+        return this.dragStart.asObservable();
+    }
+
+    get dragMoved(): Observable<ClrDragEvent> {
+        return this.dragMove.asObservable();
+    }
+
+    get dragEnded(): Observable<ClrDragEvent> {
+        return this.dragEnd.asObservable();
+    }
+
+    constructor(private renderer: Renderer2, private eventBus: ClrDragAndDropEventBus) {}
+
+    public dragDataTransfer?: ClrDragDataTransfer;
+
+    public attachDragListeners(draggableEl: Node) {
+        this.draggabelEl = draggableEl;
+        this.listeners = [
+            this.customDragEvent(this.draggabelEl, "mousedown", "mousemove", "mouseup"),
+            this.customDragEvent(this.draggabelEl, "touchstart", "touchmove", "touchend")
+        ];
+    }
+
+    public detachDragListeners() {
+        if (this.listeners) {
+            this.listeners.map(event => event());
+        }
+    }
+
+    private customDragEvent(element: Node, startOnEvent: string, moveOnEvent: string, endOnEvent: string): () => void {
+        let moveListener: () => void;
+        let endListener: () => void;
+
+        return this.renderer.listen(element, startOnEvent, () => {
+            moveListener = this.renderer.listen("document", moveOnEvent, (moveEvent: MouseEvent|TouchEvent) => {
+                moveEvent.preventDefault();
+                moveEvent.stopPropagation();
+
+                if (!this.hasDragStarted) {
+                    this.hasDragStarted = true;
+
+                    // attach drag end listener only if drag start was fired.
+                    endListener = this.renderer.listen("document", endOnEvent, (endEvent: MouseEvent|TouchEvent) => {
+                        moveListener();  // Unregistering from mouseMove or touchMove
```

Comment:
I believe you have a bug here: if you click on a draggable element, here is what happens:
1) The `mousedown` listener is called, and attaches and `mousemove` listener.
2) The user releases the mouse without moving, nothing happens because there isn't a mouseup listener yet.
3) The user starts moving their mouse, the `mousemove` listener is called and starts doing all its things.
4) All the listeners keep firing when the user moves the mouse, even though they are not dragging at all since the mouse has been released in 2).

Reply:
Addressed the request. Added in a test for this case.